components/invitation_modal/invitation_modal.tsx

```diff
@@ -0,0 +1,396 @@
+    result: defaultResultState,
+});
+
+export class InvitationModal extends React.PureComponent<Props, State> {
+    constructor(props: Props) {
+        super(props);
+
+        this.state = {
+            ...defaultState,
+            invite: {
+                ...defaultState.invite,
+                as: (!props.canAddUsers && props.canInviteGuests) ? As.GUEST : defaultState.invite.as,
+            },
+        };
+    }
+
+    handleHide = () => {
+        this.props.actions.closeModal();
+    }
+
+    toggleCustomMessage = () => {
+        this.setState((state) => ({
+            ...state,
+            invite: {
+                ...state.invite,
+                customMessage: {
+                    ...state.invite.customMessage,
+                    open: !state.invite.customMessage.open,
+                },
+            },
+        }));
+    }
+
+    setCustomMessage = (message: string) => {
+        this.setState((state) => ({
+            ...state,
+            invite: {
+                ...state.invite,
+                customMessage: {
+                    ...state.invite.customMessage,
+                    message,
+                },
+            },
+        }));
+    }
+
+    setInviteAs = (as: As) => {
+        if (this.state.invite.as !== as) {
+            this.setState((state) => ({
+                ...state,
+                invite: {
+                    ...this.state.invite,
+                    as,
+                },
+            }));
+        }
+    }
+
+    invite = async () => {
+        const inviteAs = this.state.invite.as;
+        if (inviteAs === As.MEMBER && this.props.isCloud) {
+            trackEvent('cloud_invite_users', 'click_send_invitations', {num_invitations: this.state.invite.usersEmails.length});
+        }
+        trackEvent('invite_users', 'click_invite');
+
+        const users: UserProfile[] = [];
+        const emails: string[] = [];
+        for (const userOrEmail of this.state.invite.usersEmails) {
+            if (typeof userOrEmail === 'string' && isEmail(userOrEmail)) {
+                emails.push(userOrEmail);
+            } else if (typeof userOrEmail !== 'string') {
+                users.push(userOrEmail);
+            }
+        }
+        let invites: InviteResults = {notSent: [], sent: []};
+        if (inviteAs === As.MEMBER) {
+            const result = await this.props.actions.sendMembersInvites(this.props.currentTeam.id, users, emails);
+            invites = result.data;
+        } else if (inviteAs === As.GUEST) {
+            const result = await this.props.actions.sendGuestsInvites(
+                this.props.currentTeam.id,
+                this.state.invite.inviteChannels.channels,
+                users,
+                emails,
+                this.state.invite.customMessage.open ? this.state.invite.customMessage.message : '',
+            );
+            invites = result.data;
+        }
+
+        if (this.state.invite.usersEmailsSearch !== '') {
```

Comment:
This is existing logic from porting from fullscreen modal. There, the intent was that if there was an invite that they didn't finish typing, and the clicked invite, we would show this as someone that failed to be invited, because it was never selected as a valid user or email option.

Reply:
Thanks for clarifying! wasn't familiar with the modal implmentation